consensus/istanbul/proxy/proxied_validator_engine.go

```diff
@@ -0,0 +1,614 @@
+			}
+		}
+	}
+}
+
+func (pv *proxiedValidatorEngine) sendValEnodeShareMsgs(ps *proxySet) {
+	logger := pv.logger.New("func", "sendValEnodeShareMsgs")
+
+	for _, proxy := range ps.proxiesByID {
+		if proxy.peer != nil {
+			assignedValidators := ps.getValidatorAssignments(nil, []enode.ID{proxy.ID()})
+			valAddresses := make([]common.Address, 0, len(assignedValidators))
+			for valAddress := range assignedValidators {
+				valAddresses = append(valAddresses, valAddress)
+			}
+			logger.Info("Sending val enode share msg to proxy", "proxy peer", proxy.peer, "valAddresses", common.ConvertToStringSlice(valAddresses))
+			pv.sendValEnodesShareMsg(proxy.peer, valAddresses)
+		}
+	}
+}
+
+// sendEnodeCerts will send the appropriate enode certificate to the proxies.
+func (pv *proxiedValidatorEngine) sendEnodeCerts(ps *proxySet, enodeCerts map[enode.ID]*istanbul.EnodeCertMsg) {
+	logger := pv.logger.New("func", "sendEnodeCerts")
+
+	for proxyID, proxy := range ps.proxiesByID {
+		if proxy.peer != nil && enodeCerts[proxyID] != nil {
+			// Generate message payload.  Note that these enode certs are already signed by the validator
+			payload, err := enodeCerts[proxyID].Msg.Payload()
+			if err != nil {
+				logger.Error("Error getting payload of enode certificate message", "err", err, "proxyID", proxyID)
+			}
+
+			logger.Info("Sharing enode certificate to proxy", "proxy peer", proxy.peer, "proxyID", proxyID)
+			pv.backend.Unicast(proxy.peer, payload, istanbul.EnodeCertificateMsg)
+		}
+	}
+
+}
+
+// updateValidatorAssignments will retrieve find the validator conn set diff between the current validator
+// conn set and the proxy set's validator conn set, and apply any diff to the proxy set.
+func (pv *proxiedValidatorEngine) updateValidatorAssignments(ps *proxySet) (bool, error) {
+	newVals, rmVals, err := pv.getValidatorConnSetDiff(ps.getValidators())
+	log.Trace("Proxy Handler updating validators", "newVals", common.ConvertToStringSlice(newVals), "rmVals", common.ConvertToStringSlice(rmVals), "err", err, "func", "updateValiadtors")
+	if err != nil {
+		return false, err
+	}
+
+	valsReassigned := false
+	if len(newVals) > 0 {
+		valsReassigned = valsReassigned || ps.addRemoteValidators(newVals)
+	}
+
+	if len(rmVals) > 0 {
+		valsReassigned = valsReassigned || ps.removeRemoteValidators(rmVals)
+	}
+
+	return valsReassigned, nil
+}
+
+// This function will return a diff between the current Validator Connection set and the `validators` parameter.
+func (pv *proxiedValidatorEngine) getValidatorConnSetDiff(validators []common.Address) (newVals []common.Address, rmVals []common.Address, err error) {
+	logger := pv.logger.New("func", "getValidatorConnSetDiff")
+
+	logger.Trace("Proxied validator engine retrieving validator connection set diff", "validators", common.ConvertToStringSlice(validators))
+
+	// Get the set of active and registered validators
+	newValConnSet, err := pv.backend.RetrieveValidatorConnSet(false)
```

Comment:
This call is asking for the uncached version, but it seems that the only time this function will be invoked is immediately after an epoch change. In that case, this seems unnecessary because the cache will be updated after an epoch change anyway, so the retrieved set will not be more fresh by asking for the uncached version. As a downside to using the uncached version here, this call does not benefit from work already done by other callers, and other callers will not benefit from the work done by this call.

The reason I am so persistent about the usage of this function is that we are working to reduce the amount of EVM processing we are doing, and calculating the election, which is expected to take at least 1 million, is a non-trivial amount of work. I would suggest we remove the "useCachedVersion" parameter from this function and _always_ return the cached version.

Reply:
agreed and changed.